File: backend/app/schemas/installment.py
from typing import Optional, List
from uuid import UUID
from datetime import date, datetime
from decimal import Decimal
from pydantic import BaseModel, ConfigDict, Field
from backend.app.models.installment import InstallmentStatusEnum
from backend.app.schemas.merchant import MerchantSimpleRead


class InstallmentScheduleBase(BaseModel):
    installment_index: int
    total_installments: int
    due_date: date
    principal_amount: Decimal
    interest_amount: Decimal = Field(default=Decimal("0.00"))
    total_installment_amount: Decimal
    is_billed: bool = False
    statement_id: Optional[UUID] = None


class InstallmentScheduleRead(InstallmentScheduleBase):
    id: UUID
    installment_plan_id: UUID
    created_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)


class InstallmentPlanBase(BaseModel):
    account_id: UUID
    origin_transaction_id: Optional[UUID] = None
    product_name: str
    merchant_id: Optional[UUID] = None
    start_date: date
    total_amount: Decimal
    conversion_fee: Decimal = Field(default=Decimal("0.00"))
    interest_rate_percent: Decimal = Field(default=Decimal("0.00"))
    term_months: int
    monthly_principal: Decimal
    monthly_interest: Decimal = Field(default=Decimal("0.00"))
    monthly_payment: Decimal
    remaining_balance: Decimal
    status: InstallmentStatusEnum = InstallmentStatusEnum.ACTIVE


class InstallmentPlanCreate(BaseModel):
    account_id: UUID
    product_name: str
    merchant_id: Optional[UUID] = None
    origin_transaction_id: Optional[UUID] = None
    start_date: date
    total_amount: Decimal
    conversion_fee: Decimal = Decimal("0.00")
    interest_rate_percent: Decimal = Decimal("0.00")
    term_months: int


class InstallmentPlanRead(InstallmentPlanBase):
    id: UUID
    user_id: Optional[UUID] = None
    created_at: Optional[datetime] = None
    merchant: Optional[MerchantSimpleRead] = None
    schedules: List[InstallmentScheduleRead] = []

    model_config = ConfigDict(from_attributes=True)


class EarlySettleRequest(BaseModel):
    statement_id: Optional[UUID] = None
    fee_percent: Decimal = Decimal("2.00")
    custom_fee: Optional[Decimal] = None


class EarlySettleResponse(BaseModel):
    plan_id: UUID
    product_name: str
    settled_principal: Decimal
    early_settlement_fee: Decimal
    new_status: InstallmentStatusEnum


class InstallmentForecastRead(BaseModel):
    billing_month: str
    active_plans_count: int
    total_principal_due: Decimal
    total_interest_due: Decimal
    total_monthly_payment: Decimal

File: frontend/src/types/installment.ts
import { Merchant } from "./merchant";

export type InstallmentStatus = "ACTIVE" | "COMPLETED" | "CANCELLED" | "EARLY_SETTLED";

export interface InstallmentSchedule {
  id: string;
  installment_plan_id: string;
  statement_id?: string;
  installment_index: number;
  total_installments: number;
  due_date: string;
  principal_amount: number;
  interest_amount: number;
  total_installment_amount: number;
  is_billed: boolean;
  created_at?: string;
}

export interface InstallmentPlan {
  id: string;
  account_id: string;
  origin_transaction_id?: string;
  product_name: string;
  merchant_id?: string;
  start_date: string;
  total_amount: number;
  conversion_fee: number;
  interest_rate_percent: number;
  term_months: number;
  monthly_principal: number;
  monthly_interest: number;
  monthly_payment: number;
  remaining_balance: number;
  status: InstallmentStatus;
  created_at?: string;
  merchant?: Merchant;
  schedules?: InstallmentSchedule[];
}

export interface InstallmentForecast {
  billing_month: string;
  active_plans_count: number;
  total_principal_due: number;
  total_interest_due: number;
  total_monthly_payment: number;
}

export interface EarlySettlePayload {
  statement_id?: string;
  fee_percent?: number;
  custom_fee?: number;
}

export interface EarlySettleResult {
  plan_id: string;
  product_name: string;
  settled_principal: number;
  early_settlement_fee: number;
  new_status: InstallmentStatus;
}
